{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from autogluon.tabular import TabularPredictor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load preprocessed stock datasets\n",
    "nvidia_df = pd.read_csv(\"../data/nvidia_stock_features_selected.csv\", index_col=0, parse_dates=True)\n",
    "microsoft_df = pd.read_csv(\"../data/microsoft_stock_features_selected.csv\", index_col=0, parse_dates=True)\n",
    "apple_df = pd.read_csv(\"../data/apple_stock_features_selected.csv\", index_col=0, parse_dates=True)\n",
    "\n",
    "# Set target variable (Close Price)\n",
    "target = \"Close\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def train_autogluon(stock_name, df, preset):\n",
    "    print(f\"\\nTraining AutoGluon on {stock_name}.\")\n",
    "\n",
    "    train_size = int(0.8*len(df))\n",
    "    train_data = df[:train_size]\n",
    "    test_data = df[train_size:]\n",
    "\n",
    "    predictor = TabularPredictor(label=target, eval_metric='mean_squared_error').fit(train_data, presets = preset)\n",
    "\n",
    "    predictions = predictor.predict(test_data)\n",
    "\n",
    "    from sklearn.metrics import mean_squared_error, r2_score\n",
    "    mse = mean_squared_error(test_data[target], predictions)\n",
    "    r2 = r2_score(test_data[target], predictions)\n",
    "\n",
    "    print(f\"Mean Squared Error: {mse}\")\n",
    "    print(f\"R2 Score: {r2}\")\n",
    "\n",
    "    return predictor, mse, r2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No path specified. Models will be saved in: \"AutogluonModels\\ag-20250308_165847\"\n",
      "Verbosity: 2 (Standard Logging)\n",
      "=================== System Info ===================\n",
      "AutoGluon Version:  1.1.1\n",
      "Python Version:     3.9.21\n",
      "Operating System:   Windows\n",
      "Platform Machine:   AMD64\n",
      "Platform Version:   10.0.26100\n",
      "CPU Count:          12\n",
      "Memory Avail:       4.85 GB / 15.83 GB (30.6%)\n",
      "Disk Space Avail:   649.92 GB / 935.26 GB (69.5%)\n",
      "===================================================\n",
      "Presets specified: ['medium_quality']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training AutoGluon on NVIDIA.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Beginning AutoGluon training ...\n",
      "AutoGluon will save models to \"AutogluonModels\\ag-20250308_165847\"\n",
      "Train Data Rows:    5086\n",
      "Train Data Columns: 11\n",
      "Label Column:       Close\n",
      "AutoGluon infers your prediction problem is: 'regression' (because dtype of label-column == float and many unique label-values observed).\n",
      "\tLabel info (max, min, mean, stddev): (7.16860294342041, 0.0539741218090057, 1.02595, 1.61522)\n",
      "\tIf 'regression' is not the correct problem_type, please manually specify the problem_type parameter during Predictor init (You may specify problem_type as one of: ['binary', 'multiclass', 'regression', 'quantile'])\n",
      "Problem Type:       regression\n",
      "Preprocessing data ...\n",
      "Using Feature Generators to preprocess the data ...\n",
      "Fitting AutoMLPipelineFeatureGenerator...\n",
      "\tAvailable Memory:                    4979.22 MB\n",
      "\tTrain Data (Original)  Memory Usage: 0.43 MB (0.0% of available memory)\n",
      "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
      "\tStage 1 Generators:\n",
      "\t\tFitting AsTypeFeatureGenerator...\n",
      "\tStage 2 Generators:\n",
      "\t\tFitting FillNaFeatureGenerator...\n",
      "\tStage 3 Generators:\n",
      "\t\tFitting IdentityFeatureGenerator...\n",
      "\tStage 4 Generators:\n",
      "\t\tFitting DropUniqueFeatureGenerator...\n",
      "\tStage 5 Generators:\n",
      "\t\tFitting DropDuplicatesFeatureGenerator...\n",
      "\tTypes of features in original data (raw dtype, special dtypes):\n",
      "\t\t('float', []) : 10 | ['High', 'Low', 'Open', 'SMA_20', 'EMA_20', ...]\n",
      "\t\t('int', [])   :  1 | ['Volume']\n",
      "\tTypes of features in processed data (raw dtype, special dtypes):\n",
      "\t\t('float', []) : 10 | ['High', 'Low', 'Open', 'SMA_20', 'EMA_20', ...]\n",
      "\t\t('int', [])   :  1 | ['Volume']\n",
      "\t0.0s = Fit runtime\n",
      "\t11 features in original data used to generate 11 features in processed data.\n",
      "\tTrain Data (Processed) Memory Usage: 0.43 MB (0.0% of available memory)\n",
      "Data preprocessing and feature engineering runtime = 0.06s ...\n",
      "AutoGluon will gauge predictive performance using evaluation metric: 'mean_squared_error'\n",
      "\tThis metric's sign has been flipped to adhere to being higher_is_better. The metric score can be multiplied by -1 to get the metric value.\n",
      "\tTo change this, specify the eval_metric parameter of Predictor()\n",
      "Automatically generating train/validation split with holdout_frac=0.1, Train Rows: 4577, Val Rows: 509\n",
      "User-specified model hyperparameters to be fit:\n",
      "{\n",
      "\t'NN_TORCH': {},\n",
      "\t'GBM': [{'extra_trees': True, 'ag_args': {'name_suffix': 'XT'}}, {}, 'GBMLarge'],\n",
      "\t'CAT': {},\n",
      "\t'XGB': {},\n",
      "\t'FASTAI': {},\n",
      "\t'RF': [{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression', 'quantile']}}],\n",
      "\t'XT': [{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression', 'quantile']}}],\n",
      "\t'KNN': [{'weights': 'uniform', 'ag_args': {'name_suffix': 'Unif'}}, {'weights': 'distance', 'ag_args': {'name_suffix': 'Dist'}}],\n",
      "}\n",
      "Fitting 11 L1 models ...\n",
      "Fitting model: KNeighborsUnif ...\n",
      "\t-3.1483\t = Validation score   (-mean_squared_error)\n",
      "\t0.04s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: KNeighborsDist ...\n",
      "\t-3.4648\t = Validation score   (-mean_squared_error)\n",
      "\t0.02s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: LightGBMXT ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1000]\tvalid_set's l2: 0.00187628\n",
      "[2000]\tvalid_set's l2: 0.00159674\n",
      "[3000]\tvalid_set's l2: 0.00150521\n",
      "[4000]\tvalid_set's l2: 0.00146813\n",
      "[5000]\tvalid_set's l2: 0.00144313\n",
      "[6000]\tvalid_set's l2: 0.00142394\n",
      "[7000]\tvalid_set's l2: 0.0014094\n",
      "[8000]\tvalid_set's l2: 0.00140249\n",
      "[9000]\tvalid_set's l2: 0.00139702\n",
      "[10000]\tvalid_set's l2: 0.00138193\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\t-0.0014\t = Validation score   (-mean_squared_error)\n",
      "\t9.72s\t = Training   runtime\n",
      "\t0.18s\t = Validation runtime\n",
      "Fitting model: LightGBM ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1000]\tvalid_set's l2: 0.000869477\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\t-0.0009\t = Validation score   (-mean_squared_error)\n",
      "\t2.54s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: RandomForestMSE ...\n",
      "\t-0.0009\t = Validation score   (-mean_squared_error)\n",
      "\t6.49s\t = Training   runtime\n",
      "\t0.13s\t = Validation runtime\n",
      "Fitting model: CatBoost ...\n",
      "\t-0.001\t = Validation score   (-mean_squared_error)\n",
      "\t60.74s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: ExtraTreesMSE ...\n",
      "\t-0.0008\t = Validation score   (-mean_squared_error)\n",
      "\t2.12s\t = Training   runtime\n",
      "\t0.12s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI ...\n",
      "\tWarning: Exception caused NeuralNetFastAI to fail during training... Skipping this model.\n",
      "\t\t[WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Detailed Traceback:\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1904, in _train_and_save\n",
      "    model = self._train_single(X, y, model, X_val, y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1844, in _train_single\n",
      "    model = model.fit(X=X, y=y, X_val=X_val, y_val=y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\models\\abstract\\abstract_model.py\", line 856, in fit\n",
      "    out = self._fit(**kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\fastainn\\tabular_nn_fastai.py\", line 213, in _fit\n",
      "    try_import_fastai()\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\common\\utils\\try_import.py\", line 119, in try_import_fastai\n",
      "    import autogluon.tabular.models.fastainn.imports_helper\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\fastainn\\imports_helper.py\", line 1, in <module>\n",
      "    from fastai.tabular.all import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\tabular\\all.py\", line 1, in <module>\n",
      "    from ..basics import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\basics.py\", line 1, in <module>\n",
      "    from .data.all import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\data\\all.py\", line 1, in <module>\n",
      "    from ..torch_basics import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\torch_basics.py\", line 1, in <module>\n",
      "    from torch import multiprocessing\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\__init__.py\", line 132, in <module>\n",
      "    raise err\n",
      "OSError: [WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Fitting model: XGBoost ...\n",
      "\t-0.0007\t = Validation score   (-mean_squared_error)\n",
      "\t2.46s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch ...\n",
      "\tWarning: Exception caused NeuralNetTorch to fail during training... Skipping this model.\n",
      "\t\t[WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Detailed Traceback:\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1904, in _train_and_save\n",
      "    model = self._train_single(X, y, model, X_val, y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1844, in _train_single\n",
      "    model = model.fit(X=X, y=y, X_val=X_val, y_val=y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\models\\abstract\\abstract_model.py\", line 856, in fit\n",
      "    out = self._fit(**kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\tabular_nn\\torch\\tabular_nn_torch.py\", line 147, in _fit\n",
      "    try_import_torch()\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\common\\utils\\try_import.py\", line 127, in try_import_torch\n",
      "    import torch\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\__init__.py\", line 132, in <module>\n",
      "    raise err\n",
      "OSError: [WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Fitting model: LightGBMLarge ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1000]\tvalid_set's l2: 0.000982935\n",
      "[2000]\tvalid_set's l2: 0.000981692\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\t-0.001\t = Validation score   (-mean_squared_error)\n",
      "\t16.11s\t = Training   runtime\n",
      "\t0.14s\t = Validation runtime\n",
      "Fitting model: WeightedEnsemble_L2 ...\n",
      "\tEnsemble Weights: {'XGBoost': 0.5, 'LightGBM': 0.25, 'ExtraTreesMSE': 0.25}\n",
      "\t-0.0007\t = Validation score   (-mean_squared_error)\n",
      "\t0.33s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "AutoGluon training complete, total runtime = 105.2s ... Best model: WeightedEnsemble_L2 | Estimated inference throughput: 3137.6 rows/s (509 batch size)\n",
      "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"AutogluonModels\\ag-20250308_165847\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean Squared Error: 2795.210218608676\n",
      "R2 Score: -0.7120062743793747\n"
     ]
    }
   ],
   "source": [
    "nvidia_predictor, nvidia_mse, nvidia_r2 = train_autogluon(\"NVIDIA\",nvidia_df, preset=\"medium_quality\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No path specified. Models will be saved in: \"AutogluonModels\\ag-20250308_170034\"\n",
      "Verbosity: 2 (Standard Logging)\n",
      "=================== System Info ===================\n",
      "AutoGluon Version:  1.1.1\n",
      "Python Version:     3.9.21\n",
      "Operating System:   Windows\n",
      "Platform Machine:   AMD64\n",
      "Platform Version:   10.0.26100\n",
      "CPU Count:          12\n",
      "Memory Avail:       4.60 GB / 15.83 GB (29.1%)\n",
      "Disk Space Avail:   649.63 GB / 935.26 GB (69.5%)\n",
      "===================================================\n",
      "Presets specified: ['medium_quality']\n",
      "Beginning AutoGluon training ...\n",
      "AutoGluon will save models to \"AutogluonModels\\ag-20250308_170034\"\n",
      "Train Data Rows:    7687\n",
      "Train Data Columns: 11\n",
      "Label Column:       Close\n",
      "AutoGluon infers your prediction problem is: 'regression' (because dtype of label-column == float and many unique label-values observed).\n",
      "\tLabel info (max, min, mean, stddev): (66.36734771728516, 0.1007561311125755, 16.35864, 13.79783)\n",
      "\tIf 'regression' is not the correct problem_type, please manually specify the problem_type parameter during Predictor init (You may specify problem_type as one of: ['binary', 'multiclass', 'regression', 'quantile'])\n",
      "Problem Type:       regression\n",
      "Preprocessing data ...\n",
      "Using Feature Generators to preprocess the data ...\n",
      "Fitting AutoMLPipelineFeatureGenerator...\n",
      "\tAvailable Memory:                    4712.57 MB\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training AutoGluon on Microsoft.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\tTrain Data (Original)  Memory Usage: 0.65 MB (0.0% of available memory)\n",
      "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
      "\tStage 1 Generators:\n",
      "\t\tFitting AsTypeFeatureGenerator...\n",
      "\tStage 2 Generators:\n",
      "\t\tFitting FillNaFeatureGenerator...\n",
      "\tStage 3 Generators:\n",
      "\t\tFitting IdentityFeatureGenerator...\n",
      "\tStage 4 Generators:\n",
      "\t\tFitting DropUniqueFeatureGenerator...\n",
      "\tStage 5 Generators:\n",
      "\t\tFitting DropDuplicatesFeatureGenerator...\n",
      "\tTypes of features in original data (raw dtype, special dtypes):\n",
      "\t\t('float', []) : 10 | ['High', 'Low', 'Open', 'SMA_20', 'EMA_20', ...]\n",
      "\t\t('int', [])   :  1 | ['Volume']\n",
      "\tTypes of features in processed data (raw dtype, special dtypes):\n",
      "\t\t('float', []) : 10 | ['High', 'Low', 'Open', 'SMA_20', 'EMA_20', ...]\n",
      "\t\t('int', [])   :  1 | ['Volume']\n",
      "\t0.2s = Fit runtime\n",
      "\t11 features in original data used to generate 11 features in processed data.\n",
      "\tTrain Data (Processed) Memory Usage: 0.65 MB (0.0% of available memory)\n",
      "Data preprocessing and feature engineering runtime = 0.31s ...\n",
      "AutoGluon will gauge predictive performance using evaluation metric: 'mean_squared_error'\n",
      "\tThis metric's sign has been flipped to adhere to being higher_is_better. The metric score can be multiplied by -1 to get the metric value.\n",
      "\tTo change this, specify the eval_metric parameter of Predictor()\n",
      "Automatically generating train/validation split with holdout_frac=0.1, Train Rows: 6918, Val Rows: 769\n",
      "User-specified model hyperparameters to be fit:\n",
      "{\n",
      "\t'NN_TORCH': {},\n",
      "\t'GBM': [{'extra_trees': True, 'ag_args': {'name_suffix': 'XT'}}, {}, 'GBMLarge'],\n",
      "\t'CAT': {},\n",
      "\t'XGB': {},\n",
      "\t'FASTAI': {},\n",
      "\t'RF': [{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression', 'quantile']}}],\n",
      "\t'XT': [{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression', 'quantile']}}],\n",
      "\t'KNN': [{'weights': 'uniform', 'ag_args': {'name_suffix': 'Unif'}}, {'weights': 'distance', 'ag_args': {'name_suffix': 'Dist'}}],\n",
      "}\n",
      "Fitting 11 L1 models ...\n",
      "Fitting model: KNeighborsUnif ...\n",
      "\t-181.1576\t = Validation score   (-mean_squared_error)\n",
      "\t0.07s\t = Training   runtime\n",
      "\t0.04s\t = Validation runtime\n",
      "Fitting model: KNeighborsDist ...\n",
      "\t-218.7004\t = Validation score   (-mean_squared_error)\n",
      "\t0.07s\t = Training   runtime\n",
      "\t0.05s\t = Validation runtime\n",
      "Fitting model: LightGBMXT ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1000]\tvalid_set's l2: 0.131971\n",
      "[2000]\tvalid_set's l2: 0.105593\n",
      "[3000]\tvalid_set's l2: 0.0954616\n",
      "[4000]\tvalid_set's l2: 0.0915351\n",
      "[5000]\tvalid_set's l2: 0.0881606\n",
      "[6000]\tvalid_set's l2: 0.0860153\n",
      "[7000]\tvalid_set's l2: 0.0847412\n",
      "[8000]\tvalid_set's l2: 0.0840267\n",
      "[9000]\tvalid_set's l2: 0.0830174\n",
      "[10000]\tvalid_set's l2: 0.0824454\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\t-0.0824\t = Validation score   (-mean_squared_error)\n",
      "\t17.31s\t = Training   runtime\n",
      "\t0.43s\t = Validation runtime\n",
      "Fitting model: LightGBM ...\n",
      "\t-0.0421\t = Validation score   (-mean_squared_error)\n",
      "\t0.78s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: RandomForestMSE ...\n",
      "\t-0.0295\t = Validation score   (-mean_squared_error)\n",
      "\t8.91s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "Fitting model: CatBoost ...\n",
      "\t-0.0516\t = Validation score   (-mean_squared_error)\n",
      "\t151.59s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: ExtraTreesMSE ...\n",
      "\t-0.0311\t = Validation score   (-mean_squared_error)\n",
      "\t1.87s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI ...\n",
      "\tWarning: Exception caused NeuralNetFastAI to fail during training... Skipping this model.\n",
      "\t\t[WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Detailed Traceback:\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1904, in _train_and_save\n",
      "    model = self._train_single(X, y, model, X_val, y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1844, in _train_single\n",
      "    model = model.fit(X=X, y=y, X_val=X_val, y_val=y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\models\\abstract\\abstract_model.py\", line 856, in fit\n",
      "    out = self._fit(**kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\fastainn\\tabular_nn_fastai.py\", line 213, in _fit\n",
      "    try_import_fastai()\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\common\\utils\\try_import.py\", line 119, in try_import_fastai\n",
      "    import autogluon.tabular.models.fastainn.imports_helper\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\fastainn\\imports_helper.py\", line 1, in <module>\n",
      "    from fastai.tabular.all import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\tabular\\all.py\", line 1, in <module>\n",
      "    from ..basics import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\basics.py\", line 1, in <module>\n",
      "    from .data.all import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\data\\all.py\", line 1, in <module>\n",
      "    from ..torch_basics import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\torch_basics.py\", line 1, in <module>\n",
      "    from torch import multiprocessing\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\__init__.py\", line 132, in <module>\n",
      "    raise err\n",
      "OSError: [WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Fitting model: XGBoost ...\n",
      "\t-0.0416\t = Validation score   (-mean_squared_error)\n",
      "\t1.25s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch ...\n",
      "\tWarning: Exception caused NeuralNetTorch to fail during training... Skipping this model.\n",
      "\t\t[WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Detailed Traceback:\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1904, in _train_and_save\n",
      "    model = self._train_single(X, y, model, X_val, y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1844, in _train_single\n",
      "    model = model.fit(X=X, y=y, X_val=X_val, y_val=y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\models\\abstract\\abstract_model.py\", line 856, in fit\n",
      "    out = self._fit(**kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\tabular_nn\\torch\\tabular_nn_torch.py\", line 147, in _fit\n",
      "    try_import_torch()\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\common\\utils\\try_import.py\", line 127, in try_import_torch\n",
      "    import torch\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\__init__.py\", line 132, in <module>\n",
      "    raise err\n",
      "OSError: [WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Fitting model: LightGBMLarge ...\n",
      "\t-0.0471\t = Validation score   (-mean_squared_error)\n",
      "\t2.14s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: WeightedEnsemble_L2 ...\n",
      "\tEnsemble Weights: {'RandomForestMSE': 0.636, 'ExtraTreesMSE': 0.182, 'LightGBMLarge': 0.136, 'XGBoost': 0.045}\n",
      "\t-0.0288\t = Validation score   (-mean_squared_error)\n",
      "\t0.22s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "AutoGluon training complete, total runtime = 187.81s ... Best model: WeightedEnsemble_L2 | Estimated inference throughput: 3266.7 rows/s (769 batch size)\n",
      "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"AutogluonModels\\ag-20250308_170034\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean Squared Error: 39498.35594509834\n",
      "R2 Score: -2.074891256553938\n"
     ]
    }
   ],
   "source": [
    "microsoft_predictor, microsoft_mse, microsoft_r2 = train_autogluon(\"Microsoft\",microsoft_df, preset=\"medium_quality\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No path specified. Models will be saved in: \"AutogluonModels\\ag-20250308_170343\"\n",
      "Verbosity: 2 (Standard Logging)\n",
      "=================== System Info ===================\n",
      "AutoGluon Version:  1.1.1\n",
      "Python Version:     3.9.21\n",
      "Operating System:   Windows\n",
      "Platform Machine:   AMD64\n",
      "Platform Version:   10.0.26100\n",
      "CPU Count:          12\n",
      "Memory Avail:       4.38 GB / 15.83 GB (27.7%)\n",
      "Disk Space Avail:   649.24 GB / 935.26 GB (69.4%)\n",
      "===================================================\n",
      "Presets specified: ['medium_quality']\n",
      "Beginning AutoGluon training ...\n",
      "AutoGluon will save models to \"AutogluonModels\\ag-20250308_170343\"\n",
      "Train Data Rows:    8748\n",
      "Train Data Columns: 11\n",
      "Label Column:       Close\n",
      "AutoGluon infers your prediction problem is: 'regression' (because dtype of label-column == float and many unique label-values observed).\n",
      "\tLabel info (max, min, mean, stddev): (29.72348594665528, 0.0377733446657657, 3.73959, 6.89678)\n",
      "\tIf 'regression' is not the correct problem_type, please manually specify the problem_type parameter during Predictor init (You may specify problem_type as one of: ['binary', 'multiclass', 'regression', 'quantile'])\n",
      "Problem Type:       regression\n",
      "Preprocessing data ...\n",
      "Using Feature Generators to preprocess the data ...\n",
      "Fitting AutoMLPipelineFeatureGenerator...\n",
      "\tAvailable Memory:                    4484.88 MB\n",
      "\tTrain Data (Original)  Memory Usage: 0.73 MB (0.0% of available memory)\n",
      "\tInferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.\n",
      "\tStage 1 Generators:\n",
      "\t\tFitting AsTypeFeatureGenerator...\n",
      "\tStage 2 Generators:\n",
      "\t\tFitting FillNaFeatureGenerator...\n",
      "\tStage 3 Generators:\n",
      "\t\tFitting IdentityFeatureGenerator...\n",
      "\tStage 4 Generators:\n",
      "\t\tFitting DropUniqueFeatureGenerator...\n",
      "\tStage 5 Generators:\n",
      "\t\tFitting DropDuplicatesFeatureGenerator...\n",
      "\tTypes of features in original data (raw dtype, special dtypes):\n",
      "\t\t('float', []) : 10 | ['High', 'Low', 'Open', 'SMA_20', 'EMA_20', ...]\n",
      "\t\t('int', [])   :  1 | ['Volume']\n",
      "\tTypes of features in processed data (raw dtype, special dtypes):\n",
      "\t\t('float', []) : 10 | ['High', 'Low', 'Open', 'SMA_20', 'EMA_20', ...]\n",
      "\t\t('int', [])   :  1 | ['Volume']\n",
      "\t0.1s = Fit runtime\n",
      "\t11 features in original data used to generate 11 features in processed data.\n",
      "\tTrain Data (Processed) Memory Usage: 0.73 MB (0.0% of available memory)\n",
      "Data preprocessing and feature engineering runtime = 0.12s ...\n",
      "AutoGluon will gauge predictive performance using evaluation metric: 'mean_squared_error'\n",
      "\tThis metric's sign has been flipped to adhere to being higher_is_better. The metric score can be multiplied by -1 to get the metric value.\n",
      "\tTo change this, specify the eval_metric parameter of Predictor()\n",
      "Automatically generating train/validation split with holdout_frac=0.1, Train Rows: 7873, Val Rows: 875\n",
      "User-specified model hyperparameters to be fit:\n",
      "{\n",
      "\t'NN_TORCH': {},\n",
      "\t'GBM': [{'extra_trees': True, 'ag_args': {'name_suffix': 'XT'}}, {}, 'GBMLarge'],\n",
      "\t'CAT': {},\n",
      "\t'XGB': {},\n",
      "\t'FASTAI': {},\n",
      "\t'RF': [{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression', 'quantile']}}],\n",
      "\t'XT': [{'criterion': 'gini', 'ag_args': {'name_suffix': 'Gini', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'entropy', 'ag_args': {'name_suffix': 'Entr', 'problem_types': ['binary', 'multiclass']}}, {'criterion': 'squared_error', 'ag_args': {'name_suffix': 'MSE', 'problem_types': ['regression', 'quantile']}}],\n",
      "\t'KNN': [{'weights': 'uniform', 'ag_args': {'name_suffix': 'Unif'}}, {'weights': 'distance', 'ag_args': {'name_suffix': 'Dist'}}],\n",
      "}\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Training AutoGluon on Apple.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fitting 11 L1 models ...\n",
      "Fitting model: KNeighborsUnif ...\n",
      "\t-53.8889\t = Validation score   (-mean_squared_error)\n",
      "\t0.04s\t = Training   runtime\n",
      "\t0.03s\t = Validation runtime\n",
      "Fitting model: KNeighborsDist ...\n",
      "\t-60.4619\t = Validation score   (-mean_squared_error)\n",
      "\t0.03s\t = Training   runtime\n",
      "\t0.02s\t = Validation runtime\n",
      "Fitting model: LightGBMXT ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1000]\tvalid_set's l2: 0.0192208\n",
      "[2000]\tvalid_set's l2: 0.014871\n",
      "[3000]\tvalid_set's l2: 0.0132518\n",
      "[4000]\tvalid_set's l2: 0.0123939\n",
      "[5000]\tvalid_set's l2: 0.0117961\n",
      "[6000]\tvalid_set's l2: 0.0114251\n",
      "[7000]\tvalid_set's l2: 0.0111981\n",
      "[8000]\tvalid_set's l2: 0.0110834\n",
      "[9000]\tvalid_set's l2: 0.0110558\n",
      "[10000]\tvalid_set's l2: 0.0110016\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\t-0.011\t = Validation score   (-mean_squared_error)\n",
      "\t12.65s\t = Training   runtime\n",
      "\t0.19s\t = Validation runtime\n",
      "Fitting model: LightGBM ...\n",
      "\t-0.0075\t = Validation score   (-mean_squared_error)\n",
      "\t0.77s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: RandomForestMSE ...\n",
      "\t-0.0065\t = Validation score   (-mean_squared_error)\n",
      "\t7.46s\t = Training   runtime\n",
      "\t0.11s\t = Validation runtime\n",
      "Fitting model: CatBoost ...\n",
      "\t-0.0147\t = Validation score   (-mean_squared_error)\n",
      "\t61.77s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "Fitting model: ExtraTreesMSE ...\n",
      "\t-0.0047\t = Validation score   (-mean_squared_error)\n",
      "\t1.8s\t = Training   runtime\n",
      "\t0.08s\t = Validation runtime\n",
      "Fitting model: NeuralNetFastAI ...\n",
      "\tWarning: Exception caused NeuralNetFastAI to fail during training... Skipping this model.\n",
      "\t\t[WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Detailed Traceback:\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1904, in _train_and_save\n",
      "    model = self._train_single(X, y, model, X_val, y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1844, in _train_single\n",
      "    model = model.fit(X=X, y=y, X_val=X_val, y_val=y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\models\\abstract\\abstract_model.py\", line 856, in fit\n",
      "    out = self._fit(**kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\fastainn\\tabular_nn_fastai.py\", line 213, in _fit\n",
      "    try_import_fastai()\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\common\\utils\\try_import.py\", line 119, in try_import_fastai\n",
      "    import autogluon.tabular.models.fastainn.imports_helper\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\fastainn\\imports_helper.py\", line 1, in <module>\n",
      "    from fastai.tabular.all import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\tabular\\all.py\", line 1, in <module>\n",
      "    from ..basics import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\basics.py\", line 1, in <module>\n",
      "    from .data.all import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\data\\all.py\", line 1, in <module>\n",
      "    from ..torch_basics import *\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\fastai\\torch_basics.py\", line 1, in <module>\n",
      "    from torch import multiprocessing\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\__init__.py\", line 132, in <module>\n",
      "    raise err\n",
      "OSError: [WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Fitting model: XGBoost ...\n",
      "\t-0.0095\t = Validation score   (-mean_squared_error)\n",
      "\t1.39s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: NeuralNetTorch ...\n",
      "\tWarning: Exception caused NeuralNetTorch to fail during training... Skipping this model.\n",
      "\t\t[WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Detailed Traceback:\n",
      "Traceback (most recent call last):\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1904, in _train_and_save\n",
      "    model = self._train_single(X, y, model, X_val, y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\trainer\\abstract_trainer.py\", line 1844, in _train_single\n",
      "    model = model.fit(X=X, y=y, X_val=X_val, y_val=y_val, total_resources=total_resources, **model_fit_kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\core\\models\\abstract\\abstract_model.py\", line 856, in fit\n",
      "    out = self._fit(**kwargs)\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\tabular\\models\\tabular_nn\\torch\\tabular_nn_torch.py\", line 147, in _fit\n",
      "    try_import_torch()\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\autogluon\\common\\utils\\try_import.py\", line 127, in try_import_torch\n",
      "    import torch\n",
      "  File \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\__init__.py\", line 132, in <module>\n",
      "    raise err\n",
      "OSError: [WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\athar\\anaconda3\\envs\\ds_env\\lib\\site-packages\\torch\\lib\\c10_cuda.dll\" or one of its dependencies.\n",
      "Fitting model: LightGBMLarge ...\n",
      "\t-0.008\t = Validation score   (-mean_squared_error)\n",
      "\t1.92s\t = Training   runtime\n",
      "\t0.01s\t = Validation runtime\n",
      "Fitting model: WeightedEnsemble_L2 ...\n",
      "\tEnsemble Weights: {'ExtraTreesMSE': 1.0}\n",
      "\t-0.0047\t = Validation score   (-mean_squared_error)\n",
      "\t0.11s\t = Training   runtime\n",
      "\t0.0s\t = Validation runtime\n",
      "AutoGluon training complete, total runtime = 90.85s ... Best model: WeightedEnsemble_L2 | Estimated inference throughput: 10990.0 rows/s (875 batch size)\n",
      "TabularPredictor saved. To load, use: predictor = TabularPredictor.load(\"AutogluonModels\\ag-20250308_170343\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mean Squared Error: 10334.883886711881\n",
      "R2 Score: -1.4156886849654202\n"
     ]
    }
   ],
   "source": [
    "apple_predictor, apple_mse, apple_r2 = train_autogluon(\"Apple\",apple_df, preset=\"medium_quality\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ds_autoML",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
